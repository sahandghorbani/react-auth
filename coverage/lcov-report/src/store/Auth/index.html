
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/store/Auth</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> src/store/Auth</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">28.16% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>20/71</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>0/6</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">16.66% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>3/18</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">25.75% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>17/66</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line low'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="index.ts"><a href="index.ts.html">index.ts</a></td>
	<td data-value="28.16" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 28%"></div><div class="cover-empty" style="width: 72%"></div></div>
	</td>
	<td data-value="28.16" class="pct low">28.16%</td>
	<td data-value="71" class="abs low">20/71</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="16.66" class="pct low">16.66%</td>
	<td data-value="18" class="abs low">3/18</td>
	<td data-value="25.75" class="pct low">25.75%</td>
	<td data-value="66" class="abs low">17/66</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2023-06-11T08:48:40.015Z
            </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    