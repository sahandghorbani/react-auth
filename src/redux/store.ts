// store.ts

import { configureStore } from "@reduxjs/toolkit";
import { setupListeners } from "@reduxjs/toolkit/query";
import { usersApi } from "./usersApi";
import authReducer from "./authSlice";

// Create the Redux store
export const store = configureStore({
  reducer: {
    [usersApi.reducerPath]: usersApi.reducer,
    auth: authReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(usersApi.middleware),
});

// Enable automatic cache invalidation and refetching for RTQ
setupListeners(store.dispatch);

// Export the API slice
export { usersApi };
