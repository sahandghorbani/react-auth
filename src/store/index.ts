import { configureStore } from '@reduxjs/toolkit';
import AuthSlice from './Auth/index';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';

const store = configureStore({
  reducer: {
    AuthSlice: AuthSlice,
  },
});

type RootState = ReturnType<typeof store.getState>
type AppDispatch = typeof store.dispatch

export const useAppDispatch: () => AppDispatch = useDispatch
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector

export default store;