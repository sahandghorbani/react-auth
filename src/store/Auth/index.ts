import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import axios from "axios";
import { Dispatch } from "redux";
import { AuthState, UserData } from "../../types/ITypes";

const initialState: AuthState = {
  loading: false,
  user: null,
  token: null,
  message: null,
  users: [],
};

const authSlice = createSlice({
  name: "auth",
  initialState,
  reducers: {
    setAllUsers(state, action: PayloadAction<any>) {
      state.users = action.payload;
    },
    AuthStart(state) {
      state.loading = true;
      state.message = null;
    },
    AuthSuccess(state, action: PayloadAction<any>) {
      let { user, type } = action.payload;
      state.loading = false;
      state.user = user.username;
      state.token = user.token;
      state.message = type === "exist" ? "Login successful" : "registered";
      localStorage.setItem("token", user.token);
      // console.log(state.user);
    },
    AuthFailure(state, action: PayloadAction<string>) {
      state.loading = false;
      state.message = action.payload;
    },
    logoutUser(state) {
      state.loading = false;
      state.user = null;
      state.token = null;
      state.message = "logged out";
      localStorage.removeItem("token");
    },
  },
});

export const { AuthStart, AuthSuccess, AuthFailure, setAllUsers, logoutUser } =
  authSlice.actions;

export const loginUser = (username: string, password: string) => {
  return async (dispatch: Dispatch) => {
    dispatch(AuthStart());

    try {
      const users = await fetchUsers();
      const user = findUserByUsername(users, username);
      if (user) {
        if (validatePassword(user, password)) {
          dispatch(AuthSuccess({ user, type: "exist" }));
          return true;
        } else {
          dispatch(AuthFailure("username exist but incorrect password"));
          return false;
        }
      } else {
        const newUser: UserData = {
          username,
          password,
          token: Math.random().toString(36).substring(7),
        };
        const response = await addUser(newUser);
        dispatch(AuthSuccess({ user: response, type: "new" }));
      }
    } catch (error: any) {
      dispatch(AuthFailure(error.message));
      return false;
    }
  };
};

// by refreshing the mainlayout, this function will be called again
export const fethUserByToken = (token: string) => {
  return async (dispatch: Dispatch) => {
    try {
      let user = await fetchUser(token);
      dispatch(AuthSuccess({ user, type: "exist" }));
    } catch (error: any) {
      dispatch(AuthFailure(error.message));
    }
  };
};

export const fetchAllUsers = () => {
  return async (dispatch: Dispatch) => {
    try {
      const users = await fetchUsers();
      dispatch(setAllUsers(users));
    } catch (error: any) {
      dispatch(AuthFailure(error.message));
    }
  };
};

const fetchUsers = async () => {
  const response = await axios.get("http://localhost:3001/users");
  return response.data;
};

const findUserByUsername = (users: UserData[], username: string) => {
  return users.find((user: UserData) => user.username === username);
};

const validatePassword = (user: UserData, password: string) => {
  return user.password === password;
};

const fetchUser = async (token: string) => {
  const response = await axios.get("http://localhost:3001/users");
  const users = response.data;
  const user = users.find((user: UserData) => user.token === token);
  return user;
};

const addUser = async (user: UserData) => {
  const response = await axios.post("http://localhost:3001/users", user);
  return response.data;
};

export default authSlice.reducer;
