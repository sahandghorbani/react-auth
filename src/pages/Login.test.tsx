import {  render } from "@testing-library/react";
import LoginPage from "./Login";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import store from "../store";

const setupTest = () => {
  const {getByRole , container ,queryByTestId } = render(
    <MemoryRouter>
      <Provider store={store}>
        <LoginPage />
      </Provider>
    </MemoryRouter>
  );
  return {getByRole, container ,queryByTestId }
};


describe("LoginPage", () => {
  it("should not render Snackbar at first", () => {
    const {queryByTestId} = setupTest();
    const snackbarElement = queryByTestId("the-snackbar");
    expect(snackbarElement).not.toBeInTheDocument();
  });
  it('renders snapshot LoginPage', () => {
    const {container}  = setupTest()

    expect(container.firstChild).toMatchSnapshot()
  })
});
