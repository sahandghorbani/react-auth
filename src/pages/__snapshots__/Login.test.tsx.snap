// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage renders snapshot LoginPage 1`] = `
<div
  class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
>
  <form>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              name="username"
              placeholder="Username"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
              style="padding-left: 8px;"
            >
              <legend
                class="PrivateNotchedOutline-legend-6"
                style="width: 0.01px;"
              >
                <span>
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
              style="padding-left: 8px;"
            >
              <legend
                class="PrivateNotchedOutline-legend-6"
                style="width: 0.01px;"
              >
                <span>
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Log In
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </form>
</div>
`;
